dbms/src/Interpreters/Cluster.cpp

```diff
@@ -348,23 +348,30 @@ Cluster::Cluster(const Settings & settings, const std::vector<std::vector<String
         ConnectionPoolPtrs replicas;
         replicas.reserve(current.size());
 
+        Addresses shard_local_addresses;
+
         for (const auto & replica : current)
         {
-            replicas.emplace_back(std::make_shared<ConnectionPool>(
-                settings.distributed_connections_pool_size,
-                replica.host_name, replica.port, replica.resolved_address,
-                replica.default_database, replica.user, replica.password,
-                "server", Protocol::Compression::Enable, Protocol::Encryption::Disable,
-                saturate(settings.connect_timeout_with_failover_ms, settings.limits.max_execution_time),
-                saturate(settings.receive_timeout, settings.limits.max_execution_time),
-                saturate(settings.send_timeout, settings.limits.max_execution_time)));
+            if (replica.is_local)
```

Comment:
Now function `remote` have support for going to local replicas without IPC. That's good :)

By the way, we often relied to older behaviour for better testing: just write `remote('127.0.0.1', ...)` to do "distributed" query with single ClickHouse instance (as ClickHouse will establish TCP connection to itself). Now this is not true.

But we can do more:
- we can introduce special setting to treat local replicas as remote: this setting will be useful for testing and debugging;
- also we can change 127.0.0.1 to 127.0.0.2 in tests (but this won't work if server listen to 127.0.0.1 as default configuration).